File: src/types/magicalActions/_MusicTradition.ts
import * as DB from "tsondb/schema/dsl"
import { NestedTranslationMap } from "../Locale.js"

export const MusicTraditionReference = (traditionIdentifier: DB.ReferenceIdentifier) =>
  DB.TypeAlias(import.meta.url, {
    name: traditionIdentifier.entity.name + "Reference",
    comment:
      "A reference to a music tradition with the music-tradition-specific name of the entry.",
    type: () =>
      DB.Object({
        id: DB.Required({
          comment: "The music tradition’s identifier.",
          type: traditionIdentifier,
        }),
        translations: NestedTranslationMap(
          DB.Required,
          traditionIdentifier.entity.name + "ReferenceTranslation",
          DB.Object({
            name: DB.Required({
              comment: "The music-tradition-specific name of the entry.",
              type: DB.String({ minLength: 1 }),
            }),
          }),
        ),
      }),
  })

export const MusicDuration = DB.TypeAlias(import.meta.url, {
  name: "MusicDuration",
  type: () =>
    DB.Object({
      length: DB.Required({
        type: DB.IncludeIdentifier(MusicLength),
      }),
      reusability: DB.Required({
        type: DB.IncludeIdentifier(MusicReusability),
      }),
    }),
})

const MusicLength = DB.Enum(import.meta.url, {
  name: "MusicLength",
  values: () => ({
    Long: DB.EnumCase({ type: null }),
    Short: DB.EnumCase({ type: null }),
  }),
})

const MusicReusability = DB.Enum(import.meta.url, {
  name: "MusicReusability",
  values: () => ({
    OneTime: DB.EnumCase({ type: null }),
    Sustainable: DB.EnumCase({ type: null }),
  }),
})

File: Sources/OptolithDatabaseSchema/GeneratedTypes/magicalActions/_MusicTradition.swift
//
//  _MusicTradition.swift
//  OptolithDatabaseSchema
//

import DiscriminatedEnum

/// A reference to a music tradition with the music-tradition-specific name of the entry
public struct MusicTraditionReference: EntitySubtype {
    /// The music tradition's identifier.
    public let id: Int

    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<MusicTraditionReferenceTranslation>

    public init(id: Int, translations: LocaleMap<MusicTraditionReferenceTranslation>) {
        self.id = id
        self.translations = translations
    }
}

public struct MusicTraditionReferenceTranslation: EntitySubtype {
    /// The music-tradition-specific name of the entry.
    public let name: NonEmptyString

    public init(name: NonEmptyString) {
        self.name = name
    }
}

public struct MusicDuration: EntitySubtype {
    public let length: MusicLength

    public let reusability: MusicReusability

    public init(length: MusicLength, reusability: MusicReusability) {
        self.length = length
        self.reusability = reusability
    }
}

public enum MusicLength: String, EntitySubtype {
    case long = "Long"
    case short = "Short"
}

public enum MusicReusability: String, EntitySubtype {
    case oneTime = "OneTime"
    case sustainable = "Sustainable"
}
